### PR TITLE
KAFKA-18067: Kafka Streams can leak Producer client under EOS

### PR DESCRIPTION
**Pull request by frankvicky**
JIRA: KAFKA-18067

To avoid leaking producers, we should add a 'closed` flag to `StreamProducer` indicating whether we should reset prouder.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
LGTM overall, but I'm wondering if some more unit test cases would better add to reflect the original bug? Up to @frankvicky and @ableegoldman

**Review by ableegoldman — APPROVED** [review]
LGTM! It might be kind of difficult to recreate the specific edge case in a test, but maybe you can add a short unit test to ActiveTaskCreatorTest to make sure it doesn't recreate a producer when #reinitializeProducer is called when it's closed

**Comment by frankvicky** [comment]
Hi @guozhangwang @ableegoldman 
Thanks for review.
I have add a unit test for this case, PTAL 🐧

**Review by ableegoldman — APPROVED** [review]
LGTM

**Comment by ableegoldman** [comment]
Merged to trunk. Thanks for the fix!

**Comment by mjsax** [comment]
This PR introduced a regression for EOS, and was reverted for `4.0.0` release.

https://github.com/apache/kafka/pull/19078

**Pull request merged**
